### ISSUE TITLE
new Function('...')()

### ISSUE THREAD

**Issue by juandopazo** [polyfillpolyfill/polyfill-service#357]
A few of the polyfills use this pattern like `Window = new Function('return function Window() {}')()`. What's the reason for this? Can it be avoided? It won't work in CSP enabled environments.

Ping @jonathantneal

**Comment by triblondon** [comment]
Agreed, I think this pattern should be avoided. I've seen discussions on es-shim issues about CSP problems when using this.

**Comment by JakeChampion** [comment]
Polyfills which still include `new Function()`:
[Document](https://github.com/Financial-Times/polyfill-service/blob/master/polyfills/Document/polyfill.js#L10)
[Element](https://github.com/Financial-Times/polyfill-service/blob/master/polyfills/Element/polyfill.js#L10)
[HTMLPictureElement](https://github.com/Financial-Times/polyfill-service/blob/master/polyfills/HTMLPictureElement/polyfill.js#L5)
[matchMedia](https://github.com/Financial-Times/polyfill-service/blob/master/polyfills/matchMedia/polyfill.js#L30)
[Object.create](https://github.com/Financial-Times/polyfill-service/blob/master/polyfills/Object/create/polyfill.js#L8)
[setImmediate](https://github.com/Financial-Times/polyfill-service/blob/master/polyfills/setImmediate/polyfill.js#L27)
[Window](https://github.com/Financial-Times/polyfill-service/blob/master/polyfills/Window/polyfill.js#L5)

**Comment by triblondon** [comment]
As CSP becomes more popular I think this is an increasing problem.  @jonathantneal are you able to provide any advice before we go ahead and start removing these?

**Comment by triblondon** [comment]
OK, I've dug into this a bit more.  Code like this:

```
(global.Window = global.constructor = new Function('return function Window() {}')()).prototype = this;
```

Could be written as:

```
(global.Window = global.constructor = function Window() {}).prototype = this;
```

However, I think this causes a memory leak, because it creates a closure over the parent scope, and because we're assigning it to a global, it will never be GCed.  Many of the other uses of `new Function` are in similar cases, like the constructor functions for Element, and Document.  In these cases the leak is probably not a critical problem since the enclosed scope is fairly small and very unlikely to clash with any subsequent use of the polyfilled constructor.  But there are other uses which are harder to avoid, eg `setImmediate`:

```
if (typeof handler === "function") {
    handler.apply(undefined, args);
} else {
    (new Function("" + handler))();
}
```

This is required because `setImmediate` can take a function expression as a string, just as `setTimeout` can, so to run it we have to construct the function.  The only alternative here would be to use `setTimeout` which would mean losing the optimisations that come with the setImmediate polyfill (and which are essential to support other polyfills like `Promise`.

Finally, there is Picturefill, which includes:

```
/*jshint evil:true */
try{
    cssCache[css] = new Function("e", buildStr(css))(units);
} catch(e) {}
/*jshint evil:false */
```

This is used in a complex process that converts a CSS string into parsable JavaScript and then creates a function from it.  Given that they've explictly opted out of a no-eval rule to write this I imagine there was no other practical method.

So I think my initial conclusion is: probably change the global constructors, but not much we can do about `setImmediate`, `picturefill` or future similar scenarios.

**Comment by triblondon** [comment]
In the case of the empty constructors, they definitely do capture the parent scope:

```
(function() {
  var a = 42;
  window.b = function() { console.log("hello world", a); };
}());
console.log(b());

>> "hello world"  42
```

I tried creating them using `var f = new Function(); f.name='Window'` but `name` is read-only.

**Comment by JakeChampion** [comment]
Is it possible to change `name`'s descriptor to be write only?

**Comment by triblondon** [comment]
I've spent some more time on this and I think it's actually a waste of time.  It occurs to me that in many of these cases the `new Function()` syntax is being used in polyfills that are only required by browsers that don't support CSP:
- Window
- Document
- Element
- Object.create
- matchMedia

We can commit as many CSP violations as we like in browsers that don't support CSP, so these are not a problem.

For picture, there is no feasible alternative to picturefill and there is crossover between browsers that enforce CSP and browsers that don't have picture (notably IE10-11, Edge 12, and numerous FF and Chrome, albeit all at least 10 versions old).  I've raised the issue on their repo.

I can't see any alternative for setImmediate but it's probably OK because it essentially just cascades a CSP violation from higher up the stack - if you do `setImmediate('console.log(42)')` that's just as much an eval as the underlying eval that the setImmediate polyfill does in order to implement it.  If you don't pass code to setImmediate as a string, then the setImmediate polyfill doesn't run the eval, so you shouldn't hit any CSP issues.

So in summary, the only sticky wicket here is picture, it's a third party polyfill, it's been raised upstream, so nowt to do here.
